sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala

```diff
@@ -395,18 +395,31 @@ private[sql] case class EnsureRequirements(sqlContext: SQLContext) extends Rule[
     assert(requiredChildOrderings.length == children.length)
 
     // Ensure that the operator's children satisfy their output distribution requirements:
-    children = children.zip(requiredChildDistributions).map { case (child, distribution) =>
-      if (child.outputPartitioning.satisfies(distribution)) {
+    children = children.zip(requiredChildDistributions).map {
+      case (child, distribution) if child.outputPartitioning.satisfies(distribution) =>
         child
-      } else {
+      case (child, BroadcastDistribution(m1)) =>
+        child match {
+          // The child is broadcasting the same variable: keep the child.
+          case Broadcast(m2, _) if m1 == m2 => child
```

Comment:
Should we have a `BroadcastPartitioning` as the outputPartitioning of `Broadcast`? Then can have different implementation.